packages/dds/tree/src/feature-libraries/modular-schema/comparison.ts

```diff
@@ -203,6 +213,192 @@ export function allowsRepoSuperset(
+		if (a === undefined && b === undefined) {
+			return true;
+		}
+		if (a !== undefined && b !== undefined) {
+			return a.size === b.size && [...a].every((value) => b.has(value));
+		}
+		return false;
+	};
+	const getFieldKind = (x: unknown): SchemaFactoryFieldKind | undefined => {
+		switch (x) {
+			case "Value": {
+				return "required";
+			}
+			case "Optional": {
+				return "optional";
+			}
+			case "Sequence": {
+				return "array";
+			}
+			default: {
+				return undefined;
+			}
+		}
+	};
+
+	/**
+	 * The helper function tracks discrepancies between two TreeFieldStoredSchema objects.
+	 */
+	const trackFieldDiscrepancies = (
+		originalField: TreeFieldStoredSchema,
+		updatedField: TreeFieldStoredSchema,
+		key: string,
+	): FieldIncompatibility[] => {
+		const differences: FieldIncompatibility[] = [];
+		if (!areSetsEqual(originalField.types, updatedField.types)) {
+			differences.push({
+				identifier: key,
+				mismatch: "allowedTypes",
+				view: updatedField.types ? [...updatedField.types.values()] : [],
+				stored: originalField.types ? [...originalField.types.values()] : [],
+			} satisfies AllowedTypeIncompatibility);
+		}
+		if (originalField.kind !== updatedField.kind) {
+			differences.push({
+				identifier: key,
+				mismatch: "fieldKind",
+				view: getFieldKind(updatedField.kind),
+				stored: getFieldKind(originalField.kind),
+			});
+		}
+		return differences;
+	};
+
+	// Check root schema discrepancies
+	incompatibilities.push(
+		...trackFieldDiscrepancies(original.rootFieldSchema, updated.rootFieldSchema, "root"),
+	);
+
+	// Check discrepancies in node schemas
+	const nodeKeySet = new Set<TreeNodeSchemaIdentifier>();
+	for (const [key, schema] of original.nodeSchema) {
+		nodeKeySet.add(key);
+		if (schema instanceof ObjectNodeStoredSchema) {
+			if (!updated.nodeSchema.has(key)) {
+				incompatibilities.push({
+					identifier: key,
+					mismatch: "nodeKind",
+					view: undefined,
+					stored: "object",
+				});
+			}
+			const updatedSchema = updated.nodeSchema.get(key);
+			if (updatedSchema instanceof MapNodeStoredSchema) {
+				incompatibilities.push({
+					identifier: key,
+					mismatch: "nodeKind",
+					view: "map",
+					stored: "object",
+				} satisfies NodeKindIncompatibility);
+			} else if (updatedSchema instanceof ObjectNodeStoredSchema) {
+				const differences: FieldIncompatibility[] = [];
+				const fieldKeySet = new Set<FieldKey>();
+				/**
+				 * We will track three types of differences:
+				 * 1. Fields that exist in the original schema but not in the updated schema.
+				 * 2. Fields that exist in both schemas but have different contents.
+				 * 3. Fields that exist in the updated schema but not in the original schema.
+				 *
+				 * First, the original schema is iterated to track the first two types of differences.
+				 * Then, the updated schema is iterated to find the third type.
+				 */
+				for (const [fieldKey, fieldStoredSchema] of schema.objectNodeFields) {
```

Comment:
This gets pretty deeply nested. Perhaps making separate functions for tree incompatibilities & field incompatibilities could help?